### PR TITLE
FW Rate Controller: fix saturation logic for VTOLs with differential thrust enabled

### PR DESCRIPTION
**Pull request by sfuhrer**
Context: https://github.com/PX4/PX4-Autopilot/issues/21501#issuecomment-1584754445

### Solved Problem
Actuator saturation logic in the FW rate controller was not working as intended if the actuators for a certain axis aren't in the matrix 1 (servo matrix) for VTOL. That led to the rate controller being very often disabled as the controller thought it the axis was saturated. That in turn leads to very bad rate tracking. 
It is most prominent when flying a Quad tailsitter without control surfaces, as in this case all 3 axis are affected. 

### Solution
Add logic for when to use which allocation matrix.

### Changelog Entry
For release notes:
```
Bugfix: fix actuator saturation handling for VTOLs with differential thrust in fixed-wing flight
```

### Alternatives
We need to work towards keeping this customization per vehicle in the allocation.

### Test coverage
SITL tested. 

I would port this to 1.14.

### CODE REVIEW

**Comment by MaEtUgR** [comment]
> led to the rate controller being very often disabled

You mean integrator part I assume?

We had similar problems on helicopters. The allocation now does some interfacing with the saturation logic, it works but is not ideal: https://github.com/PX4/PX4-Autopilot/blob/fc357921fd69e574a3f596be25faa8f4d9cdcea9/src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessHelicopter.cpp#L222-L254

> towards keeping this customization per vehicle in the allocation

Definitely 💯

File: src/modules/fw_rate_control/FixedwingRateControl.cpp [hunk 1]
```diff
@@ -280,13 +280,51 @@ void FixedwingRateControl::Run()
...
+							saturation_positive(i) = true;
+
+						} else if (control_allocator_status.unallocated_torque[i] < -FLT_EPSILON) {
+							saturation_negative(i) = true;
+						}
+					}
+				}
+
+				// TODO: send the unallocated value directly for better anti-windup
+				_rate_control.setSaturationStatus(saturation_positive, saturation_negative);
```
Comment (MaEtUgR) [inline src/modules/fw_rate_control/FixedwingRateControl.cpp:317]: Aren't all the saturation flags overwritten in the `_rate_control.setSaturationStatus()` call from the other matrix below 🤔

Comment (sfuhrer) [reply src/modules/fw_rate_control/FixedwingRateControl.cpp:317]: yes :grimacing: 
Guess that's what reviews are for:)
I think it's fine if we fill the same flags for both matrices and then apply them once to the controller.

File: src/modules/fw_rate_control/FixedwingRateControl.cpp [hunk 2]
```diff
@@ -280,13 +280,51 @@ void FixedwingRateControl::Run()
...
+
+			// saturation handling for axis controlled by differential thrust (VTOL only)
+			if (_param_vt_fw_difthr_en.get() > 0
+			    && _control_allocator_status_subs[0].update(&control_allocator_status_diff_thr)) {
+				Vector<bool, 3> saturation_positive;
+				Vector<bool, 3> saturation_negative;
+
+				for (size_t i = 0; i < 3; i++) {
+					if (diff_thr_en(i)) {
+						if (control_allocator_status.unallocated_torque[i] > FLT_EPSILON) {
```
Comment (MaEtUgR) [inline src/modules/fw_rate_control/FixedwingRateControl.cpp:307]:
```suggestion
						if (control_allocator_status_diff_thr.unallocated_torque[i] > FLT_EPSILON) {
```
Wasn't the goal to use the updated topic?
I'll propose a change ⏳  Just taking notes here.

**Review by MaEtUgR — APPROVED** [review]
@sfuhrer I reworked it 😇 What do you think? Please check if it really works as expected.

File: src/lib/rate_control/rate_control.cpp [hunk 3]
```diff
@@ -47,13 +47,23 @@ void RateControl::setPidGains(const Vector3f &P, const Vector3f &I, const Vector
...
+void RateControl::setSaturationStatus(const Vector3<bool> &saturation_positive,
+				      const Vector3<bool> &saturation_negative)
 {
 	_control_allocator_saturation_positive = saturation_positive;
 	_control_allocator_saturation_negative = saturation_negative;
 }
 
+void RateControl::setPositiveSaturationFlag(size_t axis, bool is_saturated)
+{
+	_control_allocator_saturation_positive(axis) = is_saturated;
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/lib/rate_control/rate_control.cpp]: @Jaeyoung-Lim You're completely right. Thanks for pointing out. I added a guard using rebase:
https://github.com/PX4/PX4-Autopilot/compare/1c6088b431f663491712c983a74d31ba54ef3eb0..687c049ec9612403c2f54ebca08bf1fd588c5848

**Review by MaEtUgR — APPROVED** [review]

**Comment by sfuhrer** [comment]
@MaEtUgR thanks a lot for the rework, much cleaner and it now even does what it was intended to do. 
I've rebased and squashed the commits, ready to go from my side.

File: src/modules/fw_rate_control/FixedwingRateControl.cpp [hunk 4]
```diff
@@ -51,6 +51,8 @@ FixedwingRateControl::FixedwingRateControl(bool vtol) :
 	/* fetch initial parameter values */
 	parameters_update();
 
+	_handle_param_vt_fw_difthr_en = param_find("VT_FW_DIFTHR_EN");
```
Comment (MaEtUgR) [inline src/modules/fw_rate_control/FixedwingRateControl.cpp:54]:
```suggestion
	_handle_param_vt_fw_difthr_en = param_find("VT_FW_DIFTHR_EN");
	parameters_update();
```
No? Othwrwise it's invalid during the initial parameter fetch.

Comment (sfuhrer) [reply src/modules/fw_rate_control/FixedwingRateControl.cpp:54]: yes sure, done.

**Review by MaEtUgR — APPROVED** [review]
👍

**Comment by sfuhrer** [comment]
CI was complaining about _param_vt_fw_difthr_en being int and not int32_t. Addressed and force pushed.

**Review by MaEtUgR — APPROVED** [review]

**Pull request merged**